validphys2/src/validphys/commondataparser.py

```diff
@@ -41,20 +83,585 @@
+    """Given the metadata defining the commondata,
+    returns a dataframe with all appropiate uncertainties
+
+    Parameters
+    ----------
+    metadata: ObservableMetaData
+        instance of ObservableMetaData defining the uncertainties to be loaded
+
+    Returns
+    -------
+    pd.DataFrame
+        a dataframe containing the uncertainties
+    """
+    all_df = []
+    for ufile in metadata.paths_uncertainties:
+        uncyaml = yaml.safe_load(ufile.read_text())
+
+        mindex = pd.MultiIndex.from_tuples(
+            [(k, v["treatment"], v["type"]) for k, v in uncyaml["definitions"].items()],
+            names=["name", "treatment", "type"],
+        )
+        # I'm guessing there will be a better way of doing this than calling  dataframe twice for the same thing?
+        final_df = pd.DataFrame(
+            pd.DataFrame(uncyaml["bins"]).values,
+            columns=mindex,
+            index=range(1, metadata.ndata + 1),
+        )
+        final_df.index.name = _INDEX_NAME
+        all_df.append(final_df)
+    return pd.concat(all_df, axis=1)
+
+
+def _parse_kinematics(metadata):
+    """Given the metadata defining the commondata,
+    returns a dataframe with the kinematic information
+
+    Parameters
+    ----------
+    metadata: ObservableMetaData
+        instance of ObservableMetaData defining the kinematics to be loaded
+
+    Returns
+    -------
+    pd.DataFrame
+        a dataframe containing the kinematics
+    """
+    kinematics_file = metadata.path_kinematics
+    kinyaml = yaml.safe_load(kinematics_file.read_text())
+
+    kin_dict = {}
+    for i, dbin in enumerate(kinyaml["bins"]):
+        bin_index = i + 1
+        # TODO: for now we are dropping min/max information since it didn't exist in the past
+        # unless the point doesn't have a mid value, in that case we need to generate it!
+        for d in dbin.values():
+            if d["mid"] is None:
+                d["mid"] = 0.5 * (d["max"] + d["min"])
+            d["min"] = None
+            d["max"] = None
+        kin_dict[bin_index] = pd.DataFrame(dbin).stack()
+
+    return pd.concat(kin_dict, axis=1, names=[_INDEX_NAME]).swaplevel(0, 1).T
+
+
+def parse_new_metadata(metadata_file, observable_name, variants=[]):
+    """Given a metadata file in the new format and the specific observable to be read
+    load and parse the metadata and select the observable.
+    If any variants are selected, apply them.
+    """
+    # Note: we are re-loading many times the same yaml file, possibly a good target for lru_cache
+    set_metadata = parse_yaml_inp(metadata_file, SetMetaData)
+    set_metadata._folder = metadata_file.parent
+
+    # Select one observable from the entire metadata
+    metadata = set_metadata.select_observable(observable_name)
+
+    # And apply variants
+    for variant in variants:
+        metadata = metadata.apply_variant(variant)
+
+    return metadata
+
+
+def parse_commondata_new(metadata):
+    """
```

Comment:
A metadata `ObservableMetaData` object has a `variants`  attribute.
How is this specified/enforced when loading a new commondata instance?

If I remember correctly there was a parse_commondata_new function in which the variant could be specified
(I think it was something like parse_commondata_new(datasetname, variants=['decorrelated']))

Reply:
When you ask for a dataset with a variant, the metadata is already the "variant" metadata. It is applied in `parse_new_metadata(metadata_file, observable_name, variants=[]):`

I'm doing it like this because (when variants are selected) it should be applied immediately so I moved it to just after the `metadata_file` is loaded.